File: group.js


function Group(host) = {
  this.people = {host.name: host};
  this.recipes = {};
  this.phase = "voting";
  this.add_person = function (person_id) {
    if (person_id in this.people) {
      throw "Person already exists";
    } else {
      var new_person = new Person(person_id);
      this.people[person_id] = new_person;
    }
  }
  this.delete_person = function (person_id) {
    if (!(person_id in this.people) {
      throw "Person does not exist";
    } else {
      delete this.people[person_id];
    }
  }
  this.get_person = function (person_id) {
    if (!(person_id in this.people)) {
      throw "Person does not exist";
    } else {
      return self.people[person_id];
    }
  }
  this.add_recipe = function (recipe_name) {
    if (this.phase == "voting") {
      var new_recipe = new Recipe(recipe_name);
      self.recipes[recipe_name] = new_recipe;
    } else {
      throw "Not in voting phase";
    }
  }
  this.get_recipe = function (recipe_name) {
    if (recipe_name in this.recipes) {
      return this.recipes[recipe_name];
    } else {
      throw "Recipe not found";
    }
  }
  this.process_vote = function (member, recipe, vote) {
    if (recipe in this.recipes) {
      var curr_recipe = self.recipes[recipe]
      curr_recipe.process_vote(member, vote)
    } else {
      throw "Cannot process vote";
    }
  }
  this.select_recipe = function () {
    var selected_recipe;
    for (key in this.recipes) {
      if (selected_recipe === undefined || this.recipes[key].number_votes() > selected_recipe.number_votes()) {
        selected_recipe = this.recipes[key];
      }
    }
    this.selected_recipe = selected_recipe;
    this.phase = "selected"
    this.notify_people()
  }
  this.process_confirmations(self) {
    this.phase = "confirmed"
    this.notify_people()
  }
  this.request_payments = function(message, recipe, amount) {
    var host_name = recipe.host.name;
    var payment_link = recipe.payment_link;
    var recipe_name = recipe.name;
    var request_message = host_name + " requests " + amount + " for " + recipe_name + ", pay at " + payment_link;
    this.notify_people(request_message, criteria)
  }
  this.notify_people = function(message, criteria) {
    for (person in this.people) {
      if (criteria(person)) {
        person.notify(message)
      }
    }
  }
  this.notify_members = function(message) {
    for (person in this.people) {
      person.notify(message);
    }
  }
}

File: group.py
import person
import recipe

class Group:
    """
    Main group class, handles
    calls between different
    objects.
    """
    def __init__(self, host):
        """
        TODO: Handle phase system.
        """
        self.people = {host.name: host} 
        self.recipes = {} 
        self.phase = "voting" 
        
    def add_person(self, person_id):
        """
        Adds a new person.
        TODO: Handle other person parameters.
        """
        if person_id in self.people:
            raise Exception("Person already exists!")
        else:
            new_person = Person(person_id)
            self.people[person_id] = new_person

    def delete_person(self, person_id):
        """
        Deletes person.
        """
        assert person_id in self.people
        del self.people[person_id] 

    def get_person(self, person_id):
        """
        Retrieves person.
        """
        assert person_id in self.people
        return self.people[person_id]

    def add_recipe(self, recipe_name):
        """
        Adds a new recipe.
        TODO: Handle other recipe parameters.
        """
        assert self.phase == "voting"
        new_recipe = Recipe(recipe_name)
        self.recipes[recipe_name] = new_recipe

    def get_recipe(self, recipe_name):
        """
        Retrieves recipe.
        """
        assert recipe_name in self.recipes
        return self.recipes[recipe_name]

    def process_vote(self, member, recipe, vote):
        """
        Processes incoming vote.
        """
        assert member in self.members and recipe in self.recipes
        curr_member = self.members[member]
        curr_recipe = self.recipes[recipe]
        curr_recipe.process_vote(member, vote)

    def select_recipe(self):
        """
        Selects recipe with highest votes, moves
        group onto next phase.
        TODO: Notify people.
        """
        selected_recipe = max(self.recipes, key=lambda x: x.number_votes())
        self.selected_recipe = selected_recipe
        self.phase = "selected"
        self.notify_people()

    def request_payments(self, message, recipe, amount):
        """
        Requests payments from all members
        who meet the criteria.
        """
        host_name = recipe.host.name
        payment_link = recipe.payment_link
        recipe_name = recipe.name
        request_message = "{} requests {} for {}, pay at {}".format(host_name, amount, recipe_name, payment_link) 
        self.notify_people(request_message, criteria)

    def notify_people(self, message, criteria):
        """
        Notifies people with message.
        TODO: Handle criteria function.
        """
        selected_people = [person for person in self.people if criteria(person)]
        for person in selected_people:
            person.notify(message)

